Scaling ortholearners using Ray

v-shaal opened the pull request:
issue : [793](https://github.com/py-why/EconML/issues/793)

-  Added Implementation of Ray based distributed parallelization to crossfit.
-  set flag use_ray = True or False to use ray implementation vs normal implementation
-  parallelized fit_nuisance via ray .
-  Added Testcases to compare ray vs regular implementation
- Current PR implementation is for DML , can be extended to other estimators using _Othrolearners as baseclass

kbattocchi reviewed (changes requested):
Overall, this looks like a great addition to the library.  However, there are a few changes that need to be addressed before it can be added.

First of all, please revert your changes to setup.cfg, merge the latest main back into your branch, and then make those changes to pyproject.toml instead - sorry that we changed this out from under you while your PR was in progress, but the package metadata has been moved there instead.

In addition to my comments on individual files, here are some other thoughts:
* To be broadly useful, these changes need to be propagated to at least the main DML subclasses, rather than just OrthoLearner, RLearner, and DML, but really ideally to everything that uses _crossfit.
* The coverage report shows that most of the new code in the _OrthoLearner class is never run by any of the tests, since you set `use_ray=False` for all of the tests that use the class directly.  Setting `use_ray=True` should fix that specific coverage issue, but consider whether additional tests for RLearner or DML would also be useful.
* This seems like a potentially very helpful feature, so it's probably worth creating a documentation page or notebook, or at the very least an FAQ entry, describing when/why/how to use it.

v-shaal commented on econml/_ortho_learner.py:
```diff
@@ -646,9 +720,30 @@ def fit(self, Y, T, *, X=None, W=None, Z=None, sample_weight=None, freq_weight=N
+                    return self._fit_nuisances(Y, T, X, W, Z, sample_weight=sample_weight, groups=groups)
+
+                # Create Ray remote jobs for parallel processing
+                self.nuiances_ref = [ray.remote(_fit_nuisances).options(**self.ray_remote_func_opt).remote(
```
Its because or limitation on ray end , remote needs to be called on method or class. I will take care of the typo

v-shaal commented on .github/workflows/ci.yml:
```diff
@@ -193,19 +193,19 @@ jobs:
+            extras: "[tf,plt,dowhy,ray]"
           - kind: treatment
             opts: '-m "treatment_featurization" -n auto'
-            extras: "[tf,plt]"
```
Fixed in latest commit

v-shaal commented on .github/workflows/ci.yml:
```diff
@@ -114,12 +114,12 @@ jobs:
-          extras: "[plt,dowhy]"
+          extras: "[plt,dowhy,ray]"
           pattern: "CustomerScenarios"
           version: '3.9'
```
Fixed in latest commit

v-shaal commented on econml/dml/dml.py:
```diff
@@ -466,7 +466,10 @@ def __init__(self, *,
-                 random_state=None):
+                 random_state=None,
+                 use_ray=False,
+                 **ray_remote_func_options
```
Fixed in latest commit

v-shaal commented:
**What have been fixed since last commit ?**

1) Fixed ci.yml extras dependencies
2) Added Description of all the added option in doc string in case of dml and rlearner
3) Addressed chaneges suggested for _ortho_learner.py
4)Removed ray.shutdown(), it can be taken care of explicitly on case to case basis .
5)Made ray_remote_func_options as explicit dictionary argument.

**What has been added ?**

1) Extended the changes to all estimators using _crossfit.
2) Added Test case to run for with_ray and without_ray for above changes
3) Added Notebook on how to use this feature.

@kbattocchi  kindly review the latest commit  and provide feedback if any !

fverac commented on econml/dml/dml.py:
```diff
@@ -692,7 +713,13 @@ def __init__(self, *,
+                 use_ray=False,
+                 ray_remote_func_options=None,
+                 ):
+        if ray_remote_func_options is None:
```
Can you move this logic to within the OrthoLearner fit function, and remove this logic from all subclass __init__ functions. That way we avoid redundant code in all of the subclass __init__ functions and maintain a scikit-learn-like API. If interested in more context, see the "Instantiation" section of this page https://scikit-learn.org/stable/developers/develop.html#apis-of-scikit-learn-objects. 

For instance, imagine a user does the following
```
est = LinearDML(use_ray=some_dict)
est.use_ray = None # user changes their mind about use_ray
est.fit(…)
```
We want the logic of converting None to an empty dict in .fit so we can allow for this kind of behavior.

v-shaal replied:
Noted make sense, I will move the redundant code, to fit function within Ortholearner

fverac commented on econml/dml/dml.py:
```diff
@@ -642,6 +657,12 @@ class LinearDML(StatsModelsCateEstimatorMixin, DML):
         by :mod:`np.random<numpy.random>`.
 
+    use_ray: bool, default False
+        Whether to use Ray to parallelize the cross-fitting step. If True, Ray must be installed.
```
Can you fix the spacing here with a new line in between the arg descriptions. Same for SparseLinearDML and KernelDML

v-shaal replied:
Noted

kbattocchi reviewed (approved):
I've broken the tests out into a new mark and I think things look good, so I'll merge once the checks pass.  Thanks for this contribution!
